doc/performance-testing/0005_stress-test-november-2020.md

```diff
@@ -0,0 +1,126 @@
+        "  File '/usr/local/lib/python3.8/site-packages/google/cloud/storage/blob.py', line 1654, in _do_multipart_upload",
+        "    response = upload.transmit(",
+        "  File '/usr/local/lib/python3.8/site-packages/google/resumable_media/requests/upload.py', line 139, in transmit",
+        "    response = _request_helpers.http_request(",
+        "  File '/usr/local/lib/python3.8/site-packages/google/resumable_media/requests/_request_helpers.py', line 136, in http_request",
+        "    return _helpers.wait_and_retry(func, RequestsMixin._get_status_code, retry_strategy)",
+        "  File '/usr/local/lib/python3.8/site-packages/google/resumable_media/_helpers.py', line 165, in wait_and_retry",
+        "    response = func()",
+        "  File '/usr/local/lib/python3.8/site-packages/google/auth/transport/requests.py', line 464, in request",
+        "    response = super(AuthorizedSession, self).request(",
+        "  File '/usr/local/lib/python3.8/site-packages/requests/sessions.py', line 530, in request",
+        "    resp = self.send(prep, **send_kwargs)",
+        "  File '/usr/local/lib/python3.8/site-packages/requests/sessions.py', line 643, in send",
+        "    r = adapter.send(request, **kwargs)",
+        "  File '/usr/local/lib/python3.8/site-packages/requests/adapters.py', line 498, in send",
+        "    raise ConnectionError(err, request=request)",
+        "requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))"
+
+## Observations
+
+- As sessions are stored in Datastore again, Redis was under much less load. This was the main contributing factor to a request rate of double the previous limit reached (74,000 rps vs 34,000 rps). 
+- Redis only reached a peak of 11% CPU usage and 6,400 connected clients.
+- The error rate was very low. The errors happened in the middle of the test, for 4 of 20 test runs, after that there were 5 test runs with no errors. This may have been due to GCS scaling issues.
+- The previous Datastore errors did not occur, this looks to have been fixed by switching from HTTP to gRPC for Datastore calls.
+- Response times remained consistent until just over half of the allocated vCPU was used (900 of 1800 vCPU). Around this time the k8s cluster node scaling limit was reached.
+- 99th percentile response times reached 1 second at 95% CPU usage.
+- The `loadbalancing.googleapis.com/https/backend_request_count` 500 metric as reported in Grafana (81 errors) didn't align with the errors reported by the load balancer (50 errors).
+
+## Recommendations
+
+- Update all environments to use threads and gRPC
+- Reduce the Redis instance size
+- Investigate GCS retries
+- Understand GCS scaling and speak to Google about the errors
+- Investigate the load profile we use to more evenly distribute writes to GCS
+- Test what happens when the application gets saturated; this could be done under lower load with a lower k8s node limit
```

Comment:
Does this mean pushing past the point that (as with this test) we see the latencies increase, until we observe actual errors?

Reply:
Yes